File: java/UMLGraph/src/org/daisy/urakawa/core/TreeNodeFactory.java
package org.daisy.urakawa.core;

import org.daisy.urakawa.WithPresentation;
import org.daisy.urakawa.exception.MethodParameterIsEmptyStringException;
import org.daisy.urakawa.exception.MethodParameterIsNullException;

/**
 * @leafInterface see {@link org.daisy.urakawa.LeafInterface}
 * @see org.daisy.urakawa.LeafInterface
 * @stereotype OptionalLeafInterface
 * @depend - Create 1 TreeNode
 */
public interface TreeNodeFactory extends WithPresentation {
	/**
	 * Creates a new node, which is not linked to the core data tree yet.
	 * 
	 * @return cannot return null.
	 */
	public TreeNode createNode();

	/**
	 * The namespace+name combination defines the key to a map that provides
	 * specific node implementation. This is used for allowing TreeNode to be
	 * deserialized in XUK format.
	 * 
	 * @param xukLocalName
	 * @param xukNamespaceURI
	 * @return can return null (in case the NS:name specification does not match
	 *         any supported node type).
	 * @tagvalue Exceptions "MethodParameterIsNull"
	 * @throws MethodParameterIsNullException
	 *             NULL method parameters are forbidden
	 */
	public TreeNode createNode(String xukLocalName, String xukNamespaceURI)
			throws MethodParameterIsNullException, MethodParameterIsEmptyStringException;
}

File: trunk/csharp/core/core/TreeNodeFactory.cs
namespace urakawa.core
{
    /// <summary>
    /// Default implementation of <see cref="TreeNodeFactory"/>.
    /// Creates <see cref="TreeNode"/>s belonging to a specific <see cref="Presentation"/>
    /// </summary>
    /// <remarks>
    /// A <see cref="TreeNodeFactory"/> can not create <see cref="TreeNode"/>s
    /// until it has been associated with a <see cref="Presentation"/> using the
    /// <see cref="WithPresentation.Presentation"/> method
    /// </remarks>
    public sealed class TreeNodeFactory : GenericFactory<TreeNode>
    {
        /// <summary>
        /// Creates a new <see cref="TreeNode"/>
        /// </summary>
        /// <returns>The new <see cref="TreeNode"/></returns>
        /// <exception cref="exception.IsNotInitializedException">
        /// Thrown when the <see cref="Presentation"/> of the 
        /// </exception>
        public TreeNode Create()
        {
            return Create<TreeNode>();
        }
    }
}